[2단계 - 로또 구현] 우기(박병욱) 미션 제출합니다.

jujubebat opened the pull request:
안녕하세요 김고래 🐳 
2단계 로또 미션 PR 드립니다!!

간단한 기능을 추가하는 것임에도 불구하고 많은 시간을 들여서 리팩토링을 진행한 거 같아요.  느낀점과 궁금한 점을 아래에 요약해서 적어보겠습니다. 어떤 점이 부족한지 어떤 것을 공부하면 좋을지 많이 조언주시면 감사하겠습니다 😃

- "간단한 기능의 추가임에도 불구하고 쉽지 않고 오래 걸린 이유가 변경에 유연하지 않은 설계구조 때문이지 않을까?" 라는 고민을 했습니다. "객체지향의 사실과 오해"를 읽으면서 객체지향이 무엇인지 대충 감은 오는데 깊은 철학적인 개념을 온전히 이해하지 못해서 실제 설계, 구현이 쉽지 않았던거 같아요. 객체지향 설계를 잘하는 사람이 되고 싶은데 아직 공부가 부족해서 쉽지 않은거 같아요 
- 이번 미션에서는 사용자가 예외적인 입력을 했을때 다시 재입력이 가능하도록 구현하는데 많은 집중을 했습니다. 그러다보니 컨트롤러 복잡해졌습니다. 그리고 checked exception을 사용해야하는 이유와 장점이 명확하게 마음에 와닿지 않아서 자바에서 제공하는 unchecked exception을 사용했습니다. 
- TDD가 쉽지 않았습니다. 1단계 미션에서는 테스트 코드를 작성하고 구현으로 넘어갔는데, 리팩토링을 하다보니 어느 순간 TDD를 잊어버리고 구현에만 집중하는 저를 발견할 수 있었어요. 그래도 구현한 뒤에 테스트 코드를 열심히 짰습니다. 실패하는 케이스 뿐만 아니라 성공하는 케이스도 확인하도록 테스트코드를 작성하려고 노력했어요. 그리고 테스트코드 메서드 네이밍 컨밴션을 구글링을 통해 발견해서 적용해봤는데, 김고래께서는 어떤 네이밍 컨밴션을 사용하는지, 현업에서는 네이밍을 선호하는지 궁금합니다!!
- 커밋 단위을 균일하고 작게 유지 하지 못했습니다. 리팩토링을 진행하다보면 많은 도메인, 컨트롤러, 뷰가 한꺼번에 순식간에 바뀌는데 이럴때마다 머리가 복잡해지니 커밋 남기는 것을 중시하지 않았던거 같아요 그리고 제가 남긴 커밋 메세지가 다른 사람이 볼때 의미가 있을지, 이해가 잘 되는지 궁금합니다. 
- 이전 리뷰에서 Ranks에서 map을 사용하라고 하셔서 적용해봤는데, 의도하신대로 작성한건지 궁금합니다!

이외에도 더 많이 질문하고 많은 생각을 공유하고싶지만 아직 배움이 부족해서 그러지 못하네요 😢 
많은 조언 부탁드립니다!! 감사합니다 ㅎㅎ

jujubebat committed:
refactor: Map을 사용하여 좀 더 효율적으로 rank들을 관리하도록 수정함

jujubebat committed:
refactor: final 키워드 추가

jujubebat committed:
refactor: 테스트코드 DisplayName을 더 의미있게 수정함

jujubebat committed:
refactor: Long을 int로 변하는 방법 변경

jujubebat committed:
style: 전체적으로 코드 포메팅 수정

jujubebat committed:
chore: 불필요한 파일 삭제

jujubebat committed:
refactor: 메서드명 의미있게 수정함

jujubebat committed:
refactor: 성공하는 테스트 코드 추가 및 코드 인덴트 정렬

jujubebat committed:
refactor: 로또 번호 개수와 중복 예외처리 로직을 2개로 나눔

jujubebat committed:
refactor: 당첨 로또 번호와 보너스 번호를 관리하는 클래스 이름 변경

jujubebat committed:
refactor: enum으로 관리하던 메세지들을 뷰 객체 내 상수로 포함

jujubebat committed:
feat: LottoMoney 클래스 구현

jujubebat committed:
refactor: 매개변수 변경

jujubebat committed:
feat: 로또 구매 개수를 관리하는 클래스 구현

jujubebat committed:
refactor: 수동 로또 구매 기능 구현
- 로또 티켓 생성자에 돈 뿐만 아니라 LottoLine list도 넣을 수 있게 변경함 이는 수동 로또를 로또 티켓에 추가할 수 있게하기 위함임
- 로또 라인 클래스에 수동, 자동 로또 구매여부를 확인할 수 있는 필드 추가

jujubebat committed:
refactor: 구매한 수동 로또수, 자동 로또수를 출력하는 기능 구현

jujubebat committed:
refactor: 오류 메세지는 예외가 발생하는 클래스에서 관리하도록 변경함
- 전체적인 코드 포멧팅 수정

jujubebat committed:
fix: 수동 구매 개수가 0일 경우 오류가 발생하여 수정함

jujubebat committed:
refactor: LottoMoney에게 메세지를 보내서 로직을 수행하도록 리팩토링함

jujubebat committed:
refactor: 상수 위치 이동에 따른 import 수정

jujubebat committed:
chore: lotto 패키지 구조 변경

jujubebat committed:
fix: 지난 주 당첨 번호와 보너스 번호가 중복될 경우 오류 수정
- 지난 주 당첨 번호와 보너스 번호가 중복될 경우 재입력 받는 로직이 제대로 동작하지 않아 수정함

jujubebat committed:
style: 코드 인덴트 수정함

jujubebat committed:
style: 불필요한 공백라인 삭제

jujubebat commented:
# [Git] Git 기본지식 - 1
## 내용
- Git에 대한 공부를 시작했다.
- 좋은 블로그를 발견해서 깃 관련 글들을 읽었다. git add, commit, pull, push에 대해 알고 있었지만 다시 공부했다.
- 아직 git으로 협업 해본적이 없는데, git으로 어떤식으로 협업을 하는지 기본적인 개념을 알게되었다. 해당 블로그 정주행후 pro git book을 통해 공부할 예정이다.
## 링크
- [https://tech.10000lab.xyz/git/how-git-is-different.html](https://tech.10000lab.xyz/git/how-git-is-different.html)
- [https://tech.10000lab.xyz/git/using-git-as-you-work.html](https://tech.10000lab.xyz/git/using-git-as-you-work.html)

jujubebat commented:
# [OOP] 캡슐화 - 2
## 내용
- "개발자가 반드시 정복해야할 객체지향과 디자인 패턴"을 통해 공부했다.
- 어떤 변경이 또다른 변경을 낳는 현상을 방지하기 위해 캡슐화를 사용할 수 있다는 것을 배웠다.
## 링크
- [개인 기술 블로그 정리글](http://byungwook.me/oop/2021/02/24/캡슐화.html)
- [https://book.naver.com/bookdb/book_detail.nhn?bid=7255217](https://book.naver.com/bookdb/book_detail.nhn?bid=7255217)
- [https://johngrib.github.io/wiki/law-of-demeter/](https://johngrib.github.io/wiki/law-of-demeter/)

jujubebat commented:
# [OOP] 다형성 - 2
## 내용
- "개발자가 반드시 정복해야할 객체지향과 디자인 패턴"을 통해 공부했다.
- 인터페이스 상속과 구현 상속을 통해 객체에 다형성을 부여할 수 있다는 것을 공부했다.
- 다형성을 지닌 한 객체는 여러 타입의 객체로 사용할 수 있고, 이는 if 분기문을 없애줄 수 있다는 것을 배웠다.
## 링크
- [개인 기술 블로그 정리글](http://byungwook.me/oop/2021/02/24/다형성.html)
- [https://book.naver.com/bookdb/book_detail.nhn?bid=7255217](https://book.naver.com/bookdb/book_detail.nhn?bid=7255217)

ep1stas1s commented on src/main/java/lotto/domain/lotto/LottoTicket.java:
```diff
@@ -9,21 +9,24 @@
         this.lottoLines = makeLottoLines(lottoLineCount);
         this.lottoLines.addAll(lottoLines);
-        validateCreatLottoLines();
     }
```
대표 생성자를 설정하고, 추가적으로 만든 생성자가 대표생성자를 사용하는 방식으로 구현하는 건 어떨까요?
현재 구조에선 요구사항이 변경되면, 모든 생성자를 모두 수정해야할 것 같네요 :)

하나의 생성자로 통합할 수 있으면 더더욱 좋구요 ^_^

ep1stas1s commented on src/main/java/lotto/domain/rank/Ranks.java:
```diff
@@ -1,33 +1,38 @@
     public float calculateProfitRate() {
         float totalWinMoney = Arrays.stream(Rank.values())
-            .mapToInt(rank -> getNumberOfRank(rank) * rank.getMoney()).sum();
+            .mapToInt(rank -> getNumberOfRank(rank) * rank.getMoney())
```
요기는 `Rank`에 메세지를 던져서 구현해도 좋을 것 같네요 :)

ep1stas1s commented on src/main/java/lotto/domain/rank/Ranks.java:
```diff
@@ -1,33 +1,38 @@
 
-    private final List<Rank> ranks;
+    private final Map<Rank, Long> rankCounts;
+    private final int size;
```
size는 꼭 필요한 필드는 아닐 것 같네요 🤔 
필요한 메서드에서 `PurchaseCount`나 `LottoMoney`를 받아 사용해도 좋아 보여요

ep1stas1s commented on src/main/java/lotto/domain/rank/Ranks.java:
```diff
@@ -1,33 +1,38 @@
     public int getNumberOfRank(Rank rank) {
-        long count = ranks.stream().filter(it -> it.equals(rank)).count();
-        return (int) count;
+        return rankCounts.getOrDefault(rank, 0L).intValue();
```
getOrDefault 👍

ep1stas1s commented on src/main/java/lotto/domain/lotto/LottoLine.java:
```diff
@@ -11,30 +9,48 @@
-        int LottoNumberMatchCount = winningLottoLine.getLottoNumberMatchCount(lottoNumbers);
-        boolean hasBonusNumber = winningLottoLine.isContainBonusLottoNumber(lottoNumbers);
+    public Rank checkLottoLine(WinningLotto winningLotto) {
+        int LottoNumberMatchCount = winningLotto.getLottoNumberMatchCount(lottoNumbers);
```
```suggestion
        int lottoNumberMatchCount = winningLotto.getLottoNumberMatchCount(lottoNumbers);
```

ep1stas1s commented on src/main/java/lotto/domain/rank/Rank.java:
```diff
@@ -11,28 +11,28 @@
     }
 
-    public static Rank getMatchedRank(int matchCount, boolean hasBonusNumber){
+    public static Rank getMatchedRank(int matchCount, boolean hasBonusNumber) {
```
메서드명을 고민해주세요ㅎㅎㅎ
https://johngrib.github.io/wiki/static-factory-method-pattern/

ep1stas1s commented on src/test/java/domain/lotto/LottoTicketTest.java:
```diff
@@ -1,20 +1,19 @@
+    @DisplayName("금액에 따른 로또 티켓이 생성된다.")
+    void testCreateLottoTicket() {
+        LottoTicket lottoTicket = new LottoTicket(new LottoMoney(100000000));
+        assertThat(lottoTicket.getLottoLineSize()).isEqualTo(100000);
```
아래와 같이 변수와 언더바를 사용하면 좀 더 가독성이 좋아질 수 있을 것 같아요ㅎㅎ
```suggestion
        int inputMoney = 100_000_000;
        LottoTicket lottoTicket = new LottoTicket(new LottoMoney(inputMoney));
        assertThat(lottoTicket.getLottoLineSize()).isEqualTo(inputMoney / LottoMoney.LOTTO_LINE_PRICE);
```

ep1stas1s commented on src/test/java/domain/lotto/LottoTicketTest.java:
```diff
@@ -1,20 +1,19 @@
-            .isInstanceOf(IllegalArgumentException.class)
-            .hasMessageContaining(LOTTO_PURCHASE_PRICE_ERROR.getMessage());
+    @DisplayName("금액에 따른 로또 티켓이 생성된다.")
+    void testCreateLottoTicket() {
```
`LottoTicket`에 빠진 테스트들이 많은 것 같네요... 😢

ep1stas1s commented on src/main/java/lotto/domain/rank/Ranks.java:
```diff
@@ -1,33 +1,38 @@
 public class Ranks {
 
-    private final List<Rank> ranks;
+    private final Map<Rank, Long> rankCounts;
```
잘 해주셨습니다ㅎㅎㅎ 👍 👍

ep1stas1s reviewed (changes requested):
이제 두 번째 미션인데요ㅎㅎㅎ 
고민하고 학습하시는 걸 보니 충분히 잘 하고 계신다고 생각합니다 💯 
코멘트에 대한 이견 또는 궁금한 점은 언제든 환영입니다^_^

ep1stas1s commented:
**설계**
설계는 어려운게 맞습니다...ㅠㅠ 유연한 설계가 한 눈에 보이면 좋겠지만, 안 보일 경우 아래와 같이 해보세요ㅎㅎ
1. 필요한 기능(A)만 만든다.
2. 비슷한 추가 기능(A')을 만든다. (쌔~한 기분을 느낀다)
3. 또 비슷한 추가 기능(A'')을 만들어야 할 때 리팩토링을 진행하고, 기능을 추가한다.

keyword. `3의 법칙` by. Don Roberts, `YAGNI`

**커밋**
잘 하고 계십니다:+1: 
리팩토링 시 도메인, 컨트롤러, 뷰가 같이 수정되는 건 충분히 발생할 수 있습니다 :)
그리고 커밋이 어렵다면 크게 `리팩토링`, `기능 추가` 요 두 개만 이라도 나누는 연습을 해보면 좋을 것 같네요

**테스트 메서드 네이밍**
  - `@DisplayName` : 최대한 정확히 조건과 결과를 명시합니다. 만약 조건이 복잡하다면 nested class를 사용하여 가독성을 높히는 방향으로 진행합니다. [참고](https://johngrib.github.io/wiki/junit5-nested/)
  - 메서드명 : 사실 `@DisplayName`에 의존하여 조금 소홀하긴한데... [요 방법](https://hilucky.tistory.com/216)이 괜찮다고 생각해요ㅎㅎ

jujubebat committed:
refactor : 로또 티켓 내부 생성자간 중복 로직들을 제거함
this 키워드를 사용하여 생성자간 중복로직을 제거함. 로또 티켓 객체 사용자를 고려하여 3가지 종류의 생성자를 만듦. 로또티켓 관련 테스트 코드가 거의 작성되지 않아서 새롭게 작성함. 테스트 완료.

jujubebat committed:
refactor : 지역변수 네이밍 케이싱 수정함

jujubebat committed:
refactor : 언더바를 사용하여 숫자 가독성을 높임

jujubebat committed:
refactor : 정적 팩토리 메서드명 수정

jujubebat commented:
# [Design Pattern] 정적 팩터리 메서드 - 3
## 내용
- 정적 팩터리 메서드의 장점에 대해서 공부했다. 정적 팩토리 메서드 명명법에 대해 정리했다.
- 이펙티브 자바 책을 참고했다.
## 링크
- [개인 기술 블로그 정리글](http://byungwook.me/design%20pattern/2021/02/26/정적-팩터리-메서드.html)
- [https://johngrib.github.io/wiki/static-factory-method-pattern/](https://johngrib.github.io/wiki/static-factory-method-pattern/)

jujubebat committed:
refactor : 수익률 계산 로직 구현 변경함 (rank에 메세지를 던져서 구현)

jujubebat committed:
style : 코드 인덴트 수정

jujubebat commented:
# [OOP] SOLID 설계 원칙 & DI와 서비스 로케이터 - 2
## 내용
- "객체지향과 디자인패턴"책을 읽었다.
- 변경에 유연한 소프트웨어를 구현하기 위한 SOLID 설계 원칙에 대해 학습했다.
- DI와 서비스 로케이터가 뭔지 알게되었다. 
## 링크
- https://book.naver.com/bookdb/book_detail.nhn?bid=7255217

jujubebat commented:
# [OOP] 캡슐화 - 2
## 내용
- "이펙티브 자바 3판"책을 읽으며 학습했다.
- 캡슐화를 위한 두 개의 규칙("public 클래스에서는 public 필드가 아닌 접근자 메서드를 사용하라", "클래스와 멤버의 접근 권한을 최소화하라")을 학습했다.

## 링크
- [개인 기술 블로그 정리 1](http://byungwook.me/java/2021/02/27/이펙티브자바_아이템15.html)
- [개인 기술 블로그 정리 2](http://byungwook.me/java/2021/02/27/이펙티브자바_아이템16.html)

ep1stas1s commented on src/main/java/lotto/domain/lotto/util/LottoMoney.java:
```diff
@@ -0,0 +1,25 @@
+package lotto.domain.lotto.util;
```
리팩토링할 때 패키지도 같이 정리해주시면 좋겠네요 :)

ep1stas1s commented on src/main/java/lotto/domain/lotto/LottoTicket.java:
```diff
@@ -1,47 +1,64 @@
+        return (int) count;
+    }
+
+    public void validateCreateLottoLines() {
```
처음 메서드를 구현할 때 private으로 설정하는 습관을 만드셔도 좋아요ㅎㅎ

ep1stas1s commented on src/test/java/domain/lotto/LottoMoneyTest.java:
```diff
@@ -0,0 +1,33 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class LottoMoneyTest {
```
실패하는 테스트도 추가하면 좋겠네요 :)

ep1stas1s reviewed (approved):
피드백 잘 적용해주셔서 감사합니다 👍 
학습하신 내용을 바탕으로 다음 미션도 잘 진행하시길 바랍니다 ^_^
고생하셨습니다~!
